proto/buf/alpha/registry/v1alpha1/image.proto

```diff
@@ -25,10 +25,90 @@ service ImageService {
   rpc GetImage(GetImageRequest) returns (GetImageResponse);
 }
 
+// ImageMask is used in GetImageRequest to specify which parts of an image are
+// requested.
+enum ImageMask {
+  IMAGE_MASK_UNSPECIFIED = 0;
+
+  // IMAGE_MASK_MESSAGES refers to ImageFile's `google.protobuf.DescriptorProto
+  // message_type` field.
+  IMAGE_MASK_MESSAGES = 1;
+
+  // IMAGE_MASK_ENUMS refers to ImageFile's `google.protobuf.EnumDescriptorProto
+  // enum_type` field.
+  IMAGE_MASK_ENUMS = 2;
+
+  // IMAGE_MASK_SERVICES refers to ImageFile's
+  // `google.protobuf.ServiceDescriptorProto service` field.
+  IMAGE_MASK_SERVICES = 3;
+
+  // IMAGE_MASK_EXTENSIONS refers to ImageFile's
+  // `google.protobuf.FieldDescriptorProto extension` field.
+  IMAGE_MASK_EXTENSIONS = 4;
+}
+
+// TypeDependencyStrategy is used to determine which referenced types to include
+// when only a subset of types is requested.
+// 
+// Below is an example proto that is referred to by the comments on the enum options below. 
+// 
+//    message Foo {
+//      optional Bar bar = 1 [(my_option).field = "buf"];
+//      extensions 2 to 3;
+//    }
+//    extend Foo {
+//      optional Baz baz = 2;
+//    }
+//    extend google.protobuf.FieldOptions {
+//      optional Qux my_option = 51234;
+//    }
+//
+enum TypeDependencyStrategy {
```

Comment:
This is based on a comment @amckinney and I had the other day on how to calculate the transitive closure of types. The minimal strategy is good enough for what I need at this time, but for a generic API it may be nice. Unless of course we decide that the 3rd strategy / "options" is always the way to go.

Reply:
I like what you're going for here, but I agree that we should probably punt on more granular control here for now.

The only thing I'd add is that we might want this enum to be _additive_ such that the user could request, `EXTENSIONS` _and_ `OPTIONS` instead of having the implicit relationship that `OPTIONS` includes `EXTENSIONS`. The `MINIMAL` enum value is always relevant, so we could model this as something like this:

```
message GetImageRequest {
  ...
  include_extensions = 7;
  include_options = 8;
  ...
}
```

Encapsulating all of this in an `enum` is nice, but I'm not sure that we'd ever need to add any other strategies here, so the explicit parameters might be simplest (since there's only two). But to be clear, we could only include the options and extensions right? Can you imagine anything else?

Reply:
Yes I think you're right. I think originally I was considering that options are always extensions on a {message,field,file,service}. Those extensions have a special meaning though and and should probably get some custom handling. Consider someone importing `descriptor.proto`, should requesting extensions cause all options in a module to be sent over?

I can't currently think of any other way to follow dependencies, 2 bools will probably do but they need to be `exclude_` for backwards compatibility. The only worry I have is the amount of bools that end up in this request. `exclude_imports` & `exclude_source_info` seem fine and are similar to the cli, does an extra `exclude_extensions`, `exclude_option` and `type` work for you as cli args too? I don't think we need a cli arg for the equivalent of `include_mask` no?

Reply:
Good call out on the extensions defined in `descriptor.proto`. I'd prefer if there wasn't any custom handling around `descriptor.proto`, but I know that we'll need to consider future proofing here. What are your thoughts? I feel like it's simplest for us to say, "You've asked for extensions and included `descriptor.proto`, so that's what you get."

I agree that shoving more `bool` parameters into the request doesn't feel great - it might be a slippery slope. But I think we can justify a `--include-mask`, `--exclude-extensions`, and `--exclude-options` flags in `buf build` for the same reasons we're justifying it here. I don't think we need to expose them on the CLI to begin with, but anything we have on `GetImage` needs to be available on `buf build` so that people can perform this operation locally rather than via an RPC.

Reply:
I think that when you include `descriptor.proto`, this is because you aim to write/send (parts of) filedescriptors. In this case I don't see what the use is of returning you the custom options of the FileDescriptorSet used to compile the application that handles the descriptor protos would be used for.

Example, consider:
```
import "google/protobuf/descriptor.proto"
message Foo {
	google.protobuf.FileDescriptorSet fds = 1;
}
// other file same module
message Bar {
	string a = 1 [(my_option) = {}];
}
extends google.protobuf.FieldOption {
	optional Baz my_option = 51234;
}
```

If you request the types needed for `Foo`, should that include `Baz`? Without special handling, I think you'd recurse  `Foo -> FileDescriptorSet -> ... -> FieldOption -> find all extensions -> Baz`.

The worry I have about a generic `--exclude-extensions/--exclude-options` is that they don't make much sense without specifying a type to filter on.

Reply:
Resolved offline: we will treat the `google.protobuf.*Options` specially, similar to how the compiler treats them specially and by allowing the language to be extended based on these options. Looking at the example above, we can't come up with a reason as to why `Baz` should be included in a minimal FileDescriptorSet that describes Foo. If we would make a Foo message that contained a FileDescriptorSet that Foo was made with, the extensions would be declared inside the `fds` field as a concrete instance of an extension in `FieldDescriptorProto extensions = 7`. There is no need to include its definition in the FileDescriptor declaring `Foo`.